### PR TITLE
Use flock instead of just checking if there is a file to prevent deadloc...

### LINKED ISSUE #13712: Lockfile of cron.php is waking up HDD
It is not possible to spin down a HDD which is used for data storage of your ownCloud installation because of the lockfile which is written every 15 minutes to the data/ directory.
Why don't you write it somewhere else inside of the ownCloud installation?

I use a timeout of 10 minutes to spindown my HDD but it is woken up 5 minutes after it goes to sleep because of the lockfile.

(I know it is possible to change it in cron.php but maybe it should be changed or at least be configurable.)

### PR DESCRIPTION
**Pull request by BernhardPosselt**
...ks when the process crashed or the unlink wasnt done

see #3221

fixes #13712

@DeepDiver1975 @butonic @jakobsack @MorrisJobke @PVince81

### CODE REVIEW

**Comment by Raydiation** [comment]
@icewind1991

**Comment by Raydiation** [comment]
@icewind1991 regarding external storage, can you configure the data directory so that the file cant be locked? If so we need to move the lock file up one directory. Any issues with this?

File: cron.php [hunk 1]
```diff
@@ -86,32 +58,41 @@ function handleUnexpectedShutdown() {
...
-
 		// Work
 		$jobList = \OC::$server->getJobList();
 		$jobs = $jobList->getAll();
 		foreach ($jobs as $job) {
 			$job->execute($jobList, $logger);
 		}
+
+		// unlock the file
+		flock($fp, LOCK_UN);
```
Comment (DeepDiver1975) [inline cron.php]: will the file be unlocked in case the process crashes?

Comment (Raydiation) [inline cron.php]: Yep, handled by the operating system

**Comment by Raydiation** [comment]
Ok, moved the lock file up to the top directory and created the file to rule out filesystem and permission issues.

File: cron.php (hunk 1 above)
Comment (Raydiation) [inline cron.php]: See http://stackoverflow.com/questions/13522273/will-flocked-file-be-unlocked-when-the-process-die-unexpectedly

**Comment by icewind1991** [comment]
> Ok, moved the lock file up to the top directory and created the file to rule out filesystem and permission issues.

That's going to run into permissions issues, the data dir is the only place we're guarantied to have write access to

**Comment by DeepDiver1975** [comment]
> Ok, moved the lock file up to the top directory and created the file to rule out filesystem and permission issues.

hmm - writing in root is bad - we shall only operate in  data dir and config dir - oc has no access any where else

**Comment by Raydiation** [comment]
@DeepDiver1975 config/cron.lock it is then?

**Comment by Raydiation** [comment]
otherwise we need to go for something like:

* localdata/
  * cron.lock
  * apps/
     * news

which would finally get the app data directory going

**Comment by DeepDiver1975** [comment]
Well - there are basically two options
- if we want to lock cross server -> use data (and hope that locking is supported on the file system)
- we want to lock on that server only -> use temp folder

**Comment by Raydiation** [comment]
temp folder it is then

**Comment by Raydiation** [comment]
The question is how does the temp folder work for webhosting?

**Comment by Raydiation** [comment]
e.g. multiple users use owncloud and we use a file like /tmp/owncloud-cron.lock, then only owncloud install could run at the same time.

**Comment by Raydiation** [comment]
Just use the instanceid?

**Comment by DeepDiver1975** [comment]
> Just use the instanceid?

the temp manager can help ....

**Comment by DeepDiver1975** [comment]
> the temp manager can help ....

no - it cannot - sorry for the noise

**Comment by Raydiation** [comment]
ok, moved it to the tmp directory with the name owncloud-server-$instanceid-cron.lock

**Comment by Raydiation** [comment]
PS: for shared hosting the tmp dir is easily changable using an env variable http://stackoverflow.com/questions/13186069/sys-get-temp-dir-in-shared-hosting-environment

It seems like a standard thing to setup when offering shared hosting, otherwise the instanceid should still give us enough room to not have conflicting lock files.

**Comment by Raydiation** [comment]
Incorporated

File: config/config.sample.php [hunk 2]
```diff
@@ -508,6 +508,16 @@
 'cron_log' => true,
 
 /**
+ * Location of the lock file for cron executions can be specified here.
+ * Default is within the tmp directory. The file is named in the following way
+ *   owncloud-server-$INSTANCEID-cron.lock
+ * where $INSTANCEID is the string specified in the instanceid field
```
Comment (PeterPablo) [inline config/config.sample.php]: missing dot at the end

File: config/config.sample.php [hunk 3]
```diff
@@ -508,6 +508,16 @@
...
+ * Because the cron lock file is accessed in regular intervals, it may prevent enabled disk drives from spinning down.
+ * A different location for this file can solve such issues
```
Comment (PeterPablo) [inline config/config.sample.php]: missing dot at end of line

**Comment by DeepDiver1975** [comment]
@Raydiation looks good - please rebase and squash - this might get jenkins back in shape - THX

**Comment by Raydiation** [comment]
Done

**Comment by DeepDiver1975** [comment]
> Done

thx

**Comment by DeepDiver1975** [comment]
:gun: :bomb: :boom: @owncloud-bot retest this please :boom: :bomb: :gun:

**Comment by DeepDiver1975** [comment]
I'm fine with this approach :+1: 

requires testing ....

**Comment by Raydiation** [comment]
It would be interesting to see if ajax cron will wrack havoc  if no json is being returned, can someone check? IMHO it makes no sense to display an "unknown exception" message, we should just log it.

**Comment by LukasReschke** [comment]
Rebase required.

**Comment by Raydiation** [comment]
rebased

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by LukasReschke** [comment]
Let's get this in for now and see if it causes problems :+1:

**Comment by jospoortvliet** [comment]
Great work guys, see how 8.0.1 will fare - I just had the locked cron job, too ;-)

Will this fix it if an ownCloud instance has been locked since ownCloud 8.0, btw?

**Comment by Raydiation** [comment]
Fix is not in 8.1 iirc

**Comment by Raydiation** [comment]
Yes it will fix those instances

**Comment by DeepDiver1975** [comment]
> Fix is not in 8.1 iirc

fix is in 8.1 - but not in 8.0.x

**Comment by Raydiation** [comment]
yep, typo, already fixed :)

**Comment by jospoortvliet** [comment]
Rocking! Now for that back port request... ;-)

**Pull request merged**